[새라] 로또 생성기 STEP2 완료

sejungkim opened the pull request:
로또 생성기 구현 완료했습니다.
리뷰 부탁드려요!

sejungkim committed:
Create new files for lotto program

sejungkim committed:
Add design content for obtaining number of lottos

sejungkim committed:
Implement of function obtaining the number of lotto

sejungkim committed:
Add design contents about checking valid input (money)

sejungkim committed:
Add function to check valid of 'money' value

sejungkim committed:
Add design contents of getting lotto numbers

sejungkim committed:
Add design contents about making list of lottos

sejungkim committed:
Complete the buyLottos function

Added the function of
- create lotto number
- create list of lottos

sejungkim committed:
Modify some of things for readability

- The money verification part
- The part of getting the number of lottos

sejungkim committed:
Add design of checking validation of lucky number

sejungkim committed:
Add lottoInfo object to global and changes accordingly

sejungkim committed:
Implement function of checking validate lucky number

sejungkim committed:
Add design of the part of getting winning result

sejungkim committed:
Create function to get winning result

- Fixed bug by removing function to make lottos list
- Instead, added while loop in buyLottos function

sejungkim committed:
Add design for winning statistic and rate of return

sejungkim committed:
Change function name and modify prize object data

- Changed to getMatchingResult from getWinningResult
- Changed name to prizeList from prize
- Changed prizeList object structure

sejungkim committed:
Add function for winning statistic

sejungkim committed:
Add function for showing winning statistic

sejungkim committed:
Add function for rate of return

crongro commented on javascript-lotto/lotto.js:
```diff
@@ -0,0 +1,164 @@
+const lottoList = [];
+
+// Buy Lottos
+function buyLottos(money, lotto = lottoInfo, lottoSetList = lottoList) {
```
저 위에 lottoList 전역변수는 없애버리고 여기서  lottoSetList = [] 
이렇게 하면 더 좋지 않을까요?

crongro commented on javascript-lotto/lotto.js:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  // Print lotto list
+  lottoSetList.forEach(lottoSet => console.log(lottoSet));
```
print는 할 수도 안할 수도 있는 옵션기능일 수 있어요.
따라서 buyLottos의 역할이 아닐 수 있죠.
따라서 buyLottos는 생성한 로또를 만들고, 
출력은 buyLottos 의 반환값을 이용해서 하는 것이 좀더 나을거 같아요.

crongro commented on javascript-lotto/lotto.js:
```diff
@@ -0,0 +1,164 @@
+'use strict';
+
+const lottoInfo = {
```
lottoInfo 데이터 구조가 변경이 가능한 부분(max나 min도 좋고)도 많고, 좋네요.
객체/배열을 잘 활용해서 구조를 잡았어요.

crongro commented on javascript-lotto/lotto.js:
```diff
@@ -0,0 +1,164 @@
+
+// Buy Lottos
+function buyLottos(money, lotto = lottoInfo, lottoSetList = lottoList) {
+  validateMoney(money, lotto.price);
```
만약 validateMoney 함수에서 오류가 나자나요.
함수를 원래 호출한 이곳에서 throw를 받아서 자세한 에러처리를 하는게 일반적이죠.
하위함수(subroutine)에서 생긴 오류를 거기서 끝내지 말고, 본인(여기서는 buylottos)이 에러처리를 완료해주는 게 일반적이죠.
어떻게 해야 throw를 받아서 에러처리를 완료할 수 있을까요?

crongro commented on javascript-lotto/lotto.js:
```diff
@@ -0,0 +1,164 @@
+  const moneySpent = lotto.price * lottoSetList.length;
+
+  let profit = 0;
+  for (let winType in winningStat) {
```
이거 for - in 문이 필요없는지 더 고민해보세요.

crongro reviewed (changes requested):
map,  foreach,  reduce등 잘 쓰네요.
함수크기가 작고 전체 코드의 가독성이 좋아요.
몇개 리뷰 남겼으니 살펴보고 수정하세요.

sejungkim committed:
Refactor the part of calculate total profit

- Use reduce method instead of for-in statement

sejungkim committed:
Add try-catch statement for two validation parts

- The part of validate input money and lucky number

sejungkim committed:
Add and create function for showing lotto list

Change over from the part of showing lotto list in 'buyLottos' function
to 'showLottoSetList' function

sejungkim commented:
리뷰 반영이요!
* 이익 계산하는 부분 for-in 대신 reduce로 변경했습니다.
* 유효성 검사하는 부분들(입력받은 돈과 당첨번호) try-catch 사용했습니다.
* buyLottos 함수에서 로또를 출력하는 부분을 빼고 showLottoSetList라는 함수를 만들어 사용했습니다.

crongro commented on javascript-lotto/lotto.js:
```diff
@@ -145,20 +151,25 @@ function showWinningStatistic(winningStat, prizeList) {
-setLuckyNumber([1, 2, 3, 4, 5, 6]);
\ No newline at end of file
+const lottoList = buyLottos(5000);
+if (lottoList) {
```
if 안에 조건상황을 좀더 구체적이면 더 좋아요.
이렇게만 처리해두면 잠재적인 문제가 좀더 있다고 봐야해요.
지금은 buyLottos 의 반환값이 무엇이냐?를 보고 이곳에 적절한 조건을 추가해야겠네요.

crongro reviewed (approved):
리뷰는 남겼지만 참고정도로 이해하시고요.
나머지 코드는 좋네요 .

merge합니다.
